### ISSUE TITLE
Deadlock when calling MicroProfileDumpFileImmediately just before MicroProfileShutdown under linux x64

### ISSUE THREAD

**Issue by Ryp** [jonasmr/microprofile#19]
For a while now I noticed this behavior, for current and past revisions of microprofile. My use case consists of profiling a program and then exiting after dumping the profile info. Here's my gdb dump of the behavior:
```
[New Thread 0x7f4aded35700 (LWP 21069)]              
^C                        
Thread 1 "reaper" received signal SIGINT, Interrupt. 
0x00007f4ae132133d in pthread_join () from /usr/lib/libpthread.so.0                                       
(gdb) info threads        
  Id   Target Id         Frame                       
* 1    Thread 0x7f4ae2bc4740 (LWP 21048) "reaper" 0x00007f4ae132133d in pthread_join ()                   
   from /usr/lib/libpthread.so.0                     
  2    Thread 0x7f4aded35700 (LWP 21069) "reaper" 0x00007f4ae132954c in __lll_lock_wait ()                
   from /usr/lib/libpthread.so.0                     
(gdb) thread apply 1 bt   

Thread 1 (Thread 0x7f4ae2bc4740 (LWP 21048)):        
#0  0x00007f4ae132133d in pthread_join () from /usr/lib/libpthread.so.0                                   
#1  0x00007f4ae200f467 in MicroProfileThreadJoin (pThread=0x7f4ae24f76b0 <g_MicroProfile+2528368>)        
    at /home/ryp/dev/Reaper/external/microprofile/microprofile.cpp:896                                    
#2  0x00007f4ae201b7e3 in MicroProfileWebServerStop ()                                                    
    at /home/ryp/dev/Reaper/external/microprofile/microprofile.cpp:4222                                   
#3  0x00007f4ae200fa69 in MicroProfileShutdown ()    
    at /home/ryp/dev/Reaper/external/microprofile/microprofile.cpp:1129                                   
#4  0x00000001000041b2 in main () at /home/ryp/dev/Reaper/src/main.cpp:55                                 
(gdb) thread apply 2 bt                                                                                   

Thread 2 (Thread 0x7f4aded35700 (LWP 21069)):        
#0  0x00007f4ae132954c in __lll_lock_wait () from /usr/lib/libpthread.so.0                                
#1  0x00007f4ae1322976 in pthread_mutex_lock () from /usr/lib/libpthread.so.0                             
#2  0x00007f4ae200f2a5 in __gthread_mutex_lock (__mutex=0x7f4ae228af60 <MicroProfileMutex()::Mutex>)      
    at /usr/include/c++/7.1.1/x86_64-pc-linux-gnu/bits/gthr-default.h:748                                 
#3  0x00007f4ae200f2f5 in __gthread_recursive_mutex_lock (                                                
    __mutex=0x7f4ae228af60 <MicroProfileMutex()::Mutex>)                                                  
    at /usr/include/c++/7.1.1/x86_64-pc-linux-gnu/bits/gthr-default.h:810                                 
#4  0x00007f4ae2024f6e in std::recursive_mutex::lock (this=0x7f4ae228af60 <MicroProfileMutex()::Mutex>)   
    at /usr/include/c++/7.1.1/mutex:107              
#5  0x00007f4ae2025324 in MicroProfileScopeLock::MicroProfileScopeLock (this=0x7f4aded34dd0, m=...)       
    at /home/ryp/dev/Reaper/external/microprofile/microprofile.cpp:1175                                   
#6  0x00007f4ae200fb55 in MicroProfileCreateThreadLog (                                                   
    pName=0x7f4ae202d090 "MicroProfileSocketSenderThread")                                                
    at /home/ryp/dev/Reaper/external/microprofile/microprofile.cpp:1186                                   
#7  0x00007f4ae200fefd in MicroProfileOnThreadCreate (                                                    
    pThreadName=0x7f4ae202d090 "MicroProfileSocketSenderThread")                                          
    at /home/ryp/dev/Reaper/external/microprofile/microprofile.cpp:1223                                   
#8  0x00007f4ae201e436 in MicroProfileSocketSenderThread ()                                               
    at /home/ryp/dev/Reaper/external/microprofile/microprofile.cpp:4628                                   
#9  0x00007f4ae1320049 in start_thread () from /usr/lib/libpthread.so.0                                   
#10 0x00007f4adfa6cf0f in clone () from /usr/lib/libc.so.6                                                
(gdb)  
```
I'll provide a minimal repro case if needed, hope this is enough to track down the bug.
The current source that I use to repro this bug can be found [here](https://github.com/Ryp/Reaper/blob/develop/src/main.cpp#L48).

**Comment by jonasmr** [comment]
Hi.

It should be fixed in 23856bfa356c8780aa3e3c42019eef6cb0ea4ae1 - I couldn't repro it, so I'd like to hear if the fix works.

**Comment by Ryp** [comment]
Fixed it for me, thanks a lot! Also thank you for your work on this awesome lib!
